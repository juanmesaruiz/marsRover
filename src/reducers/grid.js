import { ROVER_CLEAR } from '../actions/actionTypes';

const defaultState = { x: 50, y: 25 };

const grid = (state = defaultState, action) => {
  switch (action.type) {
    case ROVER_CLEAR:
      return defaultState;

    default:
      return state;
  }
};

export default grid;
