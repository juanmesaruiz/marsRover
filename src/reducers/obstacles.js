import { ROVER_CLEAR, ROVER_SET_OBSTACLES } from '../actions/actionTypes';

const defaultState = [];

const grid = (state = defaultState, action) => {
  switch (action.type) {
    case ROVER_CLEAR:
      return defaultState;

    case ROVER_SET_OBSTACLES: {
      const { obstacles } = action.payload;
      return [...state, ...obstacles];
    }

    default:
      return state;
  }
};

export default grid;
