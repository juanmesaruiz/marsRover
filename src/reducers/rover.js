import {
  ROVER_CLEAR,
  ROVER_CLEAR_POSITION_LOG,
  ROVER_NEW_INSTRUCTIONS,
  ROVER_NEW_MOVE,
  ROVER_SET_POSITION,
} from '../actions/actionTypes';

import { ROVER_DIRECTION, ROVER_MOVEMENT } from '../common/constants';

import { getRoverMovementFromCode } from '../common/helpers';
import { getNewRoverState } from './helpers/roverHelpers';

const defaultState = {
  current: null,
  direction: ROVER_DIRECTION.N,
  log: [],
};

const rover = (state = defaultState, action) => {
  switch (action.type) {
    case ROVER_CLEAR:
      return defaultState;

    case ROVER_CLEAR_POSITION_LOG:
      return { ...state, log: [] };

    case ROVER_NEW_INSTRUCTIONS: {
      const { instructions, grid, obstaclesCoordinates } = action.payload;
      const instructionsArray = [...instructions];
      let newState = { ...state };

      instructionsArray.forEach(instruction => {
        const roverMovement = ROVER_MOVEMENT[instruction.toUpperCase()];
        const { current, direction } = newState;
        if (roverMovement) {
          newState = getNewRoverState({
            current,
            direction,
            grid,
            obstaclesCoordinates,
            roverMovement,
            state: newState,
          });
        }
      });
      return newState;
    }

    case ROVER_NEW_MOVE: {
      const { code, grid, obstaclesCoordinates } = action.payload;
      const roverMovement = getRoverMovementFromCode(code);
      const { current, direction } = state;

      return getNewRoverState({
        current,
        direction,
        grid,
        obstaclesCoordinates,
        roverMovement,
        state,
      });
    }

    case ROVER_SET_POSITION: {
      const { position } = action.payload;
      const log = state.current
        ? [...state.log, state.current, position]
        : [...state.log];

      return {
        ...state,
        current: position,
        log,
      };
    }

    default:
      return state;
  }
};

export default rover;
