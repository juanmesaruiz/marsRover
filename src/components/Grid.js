import React from 'react';
import { connect } from 'react-redux';
import styled from 'styled-components';

import { getGrid, getObstacles, getRover } from '../reducers';
import { getDirectionArrow } from '../common/helpers';

const Grid = ({
  className,
  obstacles,
  grid: { x, y },
  rover: { current, direction },
}) => {
  const renderGrid = () => {
    const gridItems = [];
    const columns = x;
    const rows = y;
    let rowsI = 0;
    const obstaclesAvailable = [...obstacles];
    for (rowsI; rowsI <= rows; rowsI++) {
      let columnsI = 0;
      const rowItems = [];

      for (columnsI; columnsI <= columns; columnsI++) {
        const xCoordinate = columnsI;
        const yCoordinate = rows - rowsI;
        const id = `${xCoordinate}-${yCoordinate}`;
        let isObstacle = false;
        let isRoverPosition = false;

        if (xCoordinate === current?.x && yCoordinate === current?.y) {
          isRoverPosition = true;
        } else if (obstaclesAvailable.length) {
          const obstacleIndex = obstaclesAvailable.findIndex(
            obstacle => obstacle.x === xCoordinate && obstacle.y === yCoordinate
          );
          if (obstacleIndex !== -1) {
            isObstacle = true;
            obstaclesAvailable.splice(obstacleIndex, 1);
          }
        }

        rowItems.push(
          <div
            className={`grid-square ${
              isRoverPosition
                ? 'grid-square-rover'
                : isObstacle
                ? 'grid-square-obstacle'
                : ''
            }`}
            id={id}
            key={id}
          >
            {isRoverPosition ? (
              <span className="grid-arrow">{getDirectionArrow(direction)}</span>
            ) : (
              ''
            )}
          </div>
        );
      }

      gridItems.push(
        <div key={rowsI} className="grid-rows">
          {rowItems}
        </div>
      );
    }

    return gridItems;
  };

  return <div className={className}>{renderGrid().map(item => item)}</div>;
};

const StyledGrid = styled(Grid)`
  display: inline-block;
  margin-bottom: 15px;
  vertical-align: top;

  .grid-rows {
    line-height: 0;
    margin: 0;

    .grid-square {
      border-top: 1px solid black;
      border-right: 1px solid black;
      display: inline-block;
      height: 1vw;
      max-width: 15px;
      max-height: 15px;
      position: relative;
      width: 1vw;

      &:first-child {
        border-left: 1px solid black;
      }
      &-obstacle {
        background-color: black;
      }

      &-rover {
        background-color: green;
      }

      .grid-arrow {
        position: absolute;
        top: 50%;
        left: 50%;
        transform: translateX(-50%);
        font-size: 1vw;
      }
    }
  }

  div:last-of-type .grid-square {
    border-bottom: 1px solid black;
  }
`;

const mapStateToProps = state => ({
  obstacles: getObstacles(state),
  grid: getGrid(state),
  rover: getRover(state),
});

export default connect(mapStateToProps, {})(StyledGrid);
