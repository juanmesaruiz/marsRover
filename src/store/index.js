import { applyMiddleware, compose, createStore } from 'redux';
import thunk from 'redux-thunk';

import reducers from '../reducers';

const configureStore = () => {
  const initialState = {};
  const enhancers = [];
  const devToolsExtension = window.__REDUX_DEVTOOLS_EXTENSION__;

  if (typeof devToolsExtension === 'function') {
    enhancers.push(devToolsExtension({ name: '- ROVER MARS -' }));
  }

  const composedEnhancers = compose(applyMiddleware(thunk), ...enhancers);

  const store = createStore(reducers, initialState, composedEnhancers);

  return { store };
};

export default configureStore;
