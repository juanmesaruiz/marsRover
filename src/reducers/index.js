import { combineReducers } from 'redux';

import grid from './grid';
import obstacles from './obstacles';
import rover from './rover';

const reducers = {
  grid,
  obstacles,
  rover,
};

export const getGrid = state => state.grid;
export const getObstacles = state => state.obstacles;
export const getRover = state => state.rover;

export default combineReducers(reducers);
