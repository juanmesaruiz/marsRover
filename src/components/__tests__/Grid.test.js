import React from 'react';
import ReactDOM from 'react-dom';
import { Provider } from 'react-redux';
import { fireEvent, render } from '@testing-library/react';
import 'jest-styled-components';

import { getGrid, getObstacles, getRover } from '../../reducers';

import Grid from '../Grid';

import store from '../../store/__mocks__/mockStore';
import state from '../../store/__mocks__/mockState';

jest.mock('../../reducers');

describe('Grid test suite', () => {
  const component = (
    <Provider store={store}>
      <Grid />
    </Provider>
  );

  const setup = () => render(component);

  beforeEach(() => {
    getGrid.mockReturnValue(state.grid);
    getObstacles.mockReturnValue(state.obstacles);
    getRover.mockReturnValue(state.rover);
  });

  afterEach(jest.clearAllMocks);

  it('Renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(component, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('Should match snapshot', () => {
    const { container } = setup();
    expect(container).toMatchSnapshot();
  });
});
