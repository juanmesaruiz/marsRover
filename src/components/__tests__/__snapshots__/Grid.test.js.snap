// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grid test suite Should match snapshot 1`] = `
.c0 {
  display: inline-block;
}

.c0 .grid-rows {
  height: 16px;
  margin: 0;
}

.c0 .grid-rows .grid-square {
  border-top: 1px solid black;
  border-right: 1px solid black;
  display: inline-block;
  width: 15px;
  height: 15px;
}

.c0 .grid-rows .grid-square:first-child {
  border-left: 1px solid black;
}

.c0 .grid-rows .grid-square-obstacle {
  background-color: black;
}

.c0 .grid-rows .grid-square-rover {
  background-color: green;
}

.c0 div:last-of-type .grid-square {
  border-bottom: 1px solid black;
}

<div>
  <div
    class="c0"
  >
    <div
      class="grid-rows"
    >
      <div
        class="grid-square "
        id="0-5"
      />
      <div
        class="grid-square "
        id="1-5"
      />
      <div
        class="grid-square "
        id="2-5"
      />
      <div
        class="grid-square "
        id="3-5"
      />
      <div
        class="grid-square "
        id="4-5"
      />
      <div
        class="grid-square "
        id="5-5"
      />
      <div
        class="grid-square "
        id="6-5"
      />
      <div
        class="grid-square "
        id="7-5"
      />
      <div
        class="grid-square "
        id="8-5"
      />
      <div
        class="grid-square "
        id="9-5"
      />
      <div
        class="grid-square "
        id="10-5"
      />
    </div>
    <div
      class="grid-rows"
    >
      <div
        class="grid-square "
        id="0-4"
      />
      <div
        class="grid-square grid-square-obstacle"
        id="1-4"
      />
      <div
        class="grid-square "
        id="2-4"
      />
      <div
        class="grid-square "
        id="3-4"
      />
      <div
        class="grid-square grid-square-obstacle"
        id="4-4"
      />
      <div
        class="grid-square "
        id="5-4"
      />
      <div
        class="grid-square "
        id="6-4"
      />
      <div
        class="grid-square "
        id="7-4"
      />
      <div
        class="grid-square "
        id="8-4"
      />
      <div
        class="grid-square "
        id="9-4"
      />
      <div
        class="grid-square "
        id="10-4"
      />
    </div>
    <div
      class="grid-rows"
    >
      <div
        class="grid-square grid-square-obstacle"
        id="0-3"
      />
      <div
        class="grid-square "
        id="1-3"
      />
      <div
        class="grid-square "
        id="2-3"
      />
      <div
        class="grid-square grid-square-obstacle"
        id="3-3"
      />
      <div
        class="grid-square "
        id="4-3"
      />
      <div
        class="grid-square "
        id="5-3"
      />
      <div
        class="grid-square "
        id="6-3"
      />
      <div
        class="grid-square "
        id="7-3"
      />
      <div
        class="grid-square "
        id="8-3"
      />
      <div
        class="grid-square "
        id="9-3"
      />
      <div
        class="grid-square "
        id="10-3"
      />
    </div>
    <div
      class="grid-rows"
    >
      <div
        class="grid-square "
        id="0-2"
      />
      <div
        class="grid-square "
        id="1-2"
      />
      <div
        class="grid-square grid-square-rover"
        id="2-2"
      >
        <span>
          â†’
        </span>
      </div>
      <div
        class="grid-square "
        id="3-2"
      />
      <div
        class="grid-square "
        id="4-2"
      />
      <div
        class="grid-square "
        id="5-2"
      />
      <div
        class="grid-square "
        id="6-2"
      />
      <div
        class="grid-square "
        id="7-2"
      />
      <div
        class="grid-square "
        id="8-2"
      />
      <div
        class="grid-square "
        id="9-2"
      />
      <div
        class="grid-square "
        id="10-2"
      />
    </div>
    <div
      class="grid-rows"
    >
      <div
        class="grid-square "
        id="0-1"
      />
      <div
        class="grid-square "
        id="1-1"
      />
      <div
        class="grid-square grid-square-obstacle"
        id="2-1"
      />
      <div
        class="grid-square "
        id="3-1"
      />
      <div
        class="grid-square "
        id="4-1"
      />
      <div
        class="grid-square "
        id="5-1"
      />
      <div
        class="grid-square "
        id="6-1"
      />
      <div
        class="grid-square "
        id="7-1"
      />
      <div
        class="grid-square "
        id="8-1"
      />
      <div
        class="grid-square "
        id="9-1"
      />
      <div
        class="grid-square "
        id="10-1"
      />
    </div>
    <div
      class="grid-rows"
    >
      <div
        class="grid-square "
        id="0-0"
      />
      <div
        class="grid-square "
        id="1-0"
      />
      <div
        class="grid-square "
        id="2-0"
      />
      <div
        class="grid-square "
        id="3-0"
      />
      <div
        class="grid-square "
        id="4-0"
      />
      <div
        class="grid-square "
        id="5-0"
      />
      <div
        class="grid-square "
        id="6-0"
      />
      <div
        class="grid-square "
        id="7-0"
      />
      <div
        class="grid-square "
        id="8-0"
      />
      <div
        class="grid-square "
        id="9-0"
      />
      <div
        class="grid-square "
        id="10-0"
      />
    </div>
  </div>
</div>
`;
